lib/TBDGen/TBDGen.cpp

```diff
@@ -636,8 +636,17 @@ void TBDGenVisitor::visitAbstractFunctionDecl(AbstractFunctionDecl *AFD) {
 
   addSymbol(SILDeclRef(AFD));
 
+  // Special case of a dynamic replacement of an @objc dynamic method inside of
+  // a generic class. We emit the objective c entry point in this case.
+  if (AFD->isObjCDynamicInGenericClass() && AFD->getDynamicallyReplacedDecl()) {
+    addSymbol(SILDeclRef(AFD, /*foreign*/true));
```

Comment:
I'm surprised that we need this. Is the objc thunk ever meant to have public linkage? If not, it doesn't have to appear in the TBD file.

Reply:
ObjC thunks have public SIL linkage. IRGen changes that linkage when it generates objective c metadata. Since we don't generate metadata for this function it stays public.

Reply:
https://github.com/apple/swift/blob/master/lib/IRGen/GenObjC.cpp#L930

Reply:
The thunk should not be IRGen-public, even in this case, because it's not part of the ABI. We don't want clients to call it directly do we?

Reply:
Perhaps it should be handled just like class methods, where instead of changing the linkage of the IRGen definition after emitting it, we compute the IRGen linkage differently.

Reply:
I'll change the SIL linkage of the SIL function. I don't see why the SILLinkage for the ObjC entry point could not be private to begin with.

Reply:
Just setting the SIL linkage of the objc entry point to private means it will get stripped out by dead function elimination.

What I was proposing is allowing the SIL linkage to be public, but instead of IRGen deciding that it's really IRGen private when it emits the metadata, we instead set a flag on the SILFunction, which together with the SIL linkage is used to compute the IRGen linkage.

We use a similar technique for class methods, where the SIL linkage corresponds to the AST access level, but the IRGen linkage is computed as follows:

- If the class is fragile and public, the linkage is public, even if the method is private, because subclasses need to emit references to the method in the vtable
- If the class is resilient and public, the linkage is private even if the method is public, because we want methods of a resilient class to always be called via the dispatch thunk and not referenced directly; that is, the method's non-dispatching entry point is not ABI, even if the method itself is public in the AST.

Reply:
It is totally okay for the objective c entry point to be eliminated. There is no way to call it. (We can’t add it to the objective c category for the generic class) We just need the native entry point for replacement.

Reply:
Why bother emitting the Objective-C entry point at all then?

Reply:
Hmm, why indeed.